pipeline_steps/common.groovy

```diff
@@ -1969,3 +1974,126 @@ void updateStringParam(String name, String value, String description){
     ParametersAction newAction = existingAction.createUpdated(updatedParam)
     currentBuild.rawBuild.addOrReplaceAction(newAction)
 }
+
+/**
+* Get the predicted duration for the current build in milliseconds.
+* Prediction is based on up to 6 recent builds.
+* See: https://javadoc.jenkins-ci.org/hudson/model/Job.html#getEstimatedDurationCandidates--
+* NonCPS because build is not serialisable
+*/
+@NonCPS
+Duration getPredictedDuration(){
+  Run build = currentBuild.rawBuild
+  Job job = build.getParent()
+  return Duration.ofMillis(job.getEstimatedDuration())
+}
+
+/**
+* Get the predicted end time for this build, based on the time it moved out of the queue
+* and its predicted duration.
+* NonCPS because build is not serialisable
+*/
+@NonCPS
+LocalDateTime getPredictedCompletionTime(){
+  Run build = currentBuild.rawBuild
+  Long startTimeInSeconds = (build.getStartTimeInMillis()/1000).round(MathContext.DECIMAL32).longValueExact()
+  LocalDateTime startTime = LocalDateTime.ofEpochSecond(startTimeInSeconds, 0, ZoneOffset.UTC)
+  return startTime.plus(getPredictedDuration())
+}
+
+/**
+* Get date object representing the start of the next maintenance window
+*/
+LocalDateTime getNextMaintenanceWindowStart(){
+  // set start of maintenance window
+
+  DayOfWeek maintDay = DayOfWeek.MONDAY
+  Integer maintHour = 10
```

Comment:
What happens if the maintenance takes more than an hour? Have you written the process down for how this will work? Are there manual step to prevent and allow new builds?

Reply:
The maintenance window is 2 hours, and we'll have to enable quiet mode manually.

Reply:
I looked into creating a job for, this but it seems quiet mode currently has problems with pipeline: https://issues.jenkins-ci.org/browse/JENKINS-34256